core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+        }
+
+        let nullable = if required {
+            ColumnNullability::NoNulls
+        } else {
+            ColumnNullability::Nullable
+        };
+
+        // Case 3: field is scalar schema
+        if field_schema.bson_type.is_some() {
+            return Ok(if field_schema.bson_type == Some(BsonTypeName::Null) {
+                ColumnNullability::Nullable
+            } else {
+                nullable
+            });
+        }
+
+        // Case 4: field is AnyOf schema
+        if let Some(any_of) = self.any_of.as_ref() {
+            for any_of_schema in any_of {
+                if any_of_schema.bson_type == Some(BsonTypeName::Null) {
+                    return Ok(ColumnNullability::Nullable);
+                }
+            }
+        } else {
+            return Err(Error::InvalidResultSetJsonSchema);
+        }
+
+        Ok(nullable)
+    }
+
+    fn is_any(&self) -> bool {
+        return self.bson_type.is_none()
+            && self.properties.is_none()
+            && self.any_of.is_none()
+            && self.required.is_none()
+            && self.items.is_none()
+            && !self.additional_properties;
+    }
+}
+
+// Converts a deserialized json_schema::Schema into a SimplifiedJsonSchema. The
+// SimplifiedJsonSchema instance is semantically equivalent to the base schema,
+// but bson_type has to be a single type otherwise the types will get pushed
+// down in the any_of list.
+impl From<json_schema::Schema> for SimplifiedJsonSchema {
+    fn from(schema: Schema) -> Self {
+        let mut properties = schema.properties.map(|properties| {
+            properties
+                .into_iter()
+                .map(|(field_name, field_schema)| (field_name, field_schema.into()))
+                .collect::<HashMap<String, SimplifiedJsonSchema>>()
+        });
+
+        let mut any_of = schema.any_of.map(|any_of| {
+            any_of
+                .into_iter()
+                .map(|any_of_schema| any_of_schema.into())
+                .collect::<Vec<SimplifiedJsonSchema>>()
+        });
+
+        let mut required = schema
+            .required
+            .map(|required| required.into_iter().collect::<BTreeSet<String>>());
+
+        let mut items = schema.items.map(|items| {
+            match items {
+                // The single-schema variant of the `items`
+                // field constrains all elements of the array.
+                Items::Single(items_schema) => Box::new(SimplifiedJsonSchema::from(*items_schema)), // actually convert
+                // The multiple-schema variant of the `items` field only
+                // asserts the schemas for the array items at specified
+                // indexes, and imposes no constraint on items at larger
+                // indexes. As such, the only schema that can describe all
+                // elements of the array is `Any`.
+                Items::Multiple(items_schemas) => Box::new(SimplifiedJsonSchema::ANY),
```

Comment:
This is the standard practice we have across our repos. In fact, this comment is copied directly from mongosql-rs. The same idea is done in the JDBC driver's MongoJsonSchema class when converting a Multiple bson_type to an any_of set.